### ISSUE TITLE
BUG: failed git detection (despite it being on PATH) on windows

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/conda-smithy#2185]
With 3.45.0, I get

```
ImportError: Failed to initialize: Bad git executable.
The git executable must be specified in one of the following ways:
    - be included in your $PATH
    - be set via $GIT_PYTHON_GIT_EXECUTABLE
    - explicitly set via git.refresh(<full-path-to-git-executable>)

All git commands will error until this is rectified.
```

It's definitely on my path though
```
>where git
C:\Users\[xxx]\.pixi\envs\git\Library\bin\git.exe
C:\Users\[xxx]\.pixi\bin\git.bat
```

**Comment by h-vetinari** [comment]
With `set "GIT_PYTHON_GIT_EXECUTABLE=C:\Users\[xxx]\.pixi\envs\git\Library\bin\git.exe"`, things did progress past the initial barrier, but then fail with
```
Traceback (most recent call last):
  File "E:\miniforge\envs\builder\Lib\site-packages\git\cmd.py", line 1262, in execute
    proc = safer_popen(
           ^^^^^^^^^^^^
  File "E:\miniforge\envs\builder\Lib\site-packages\git\cmd.py", line 281, in _safer_popen_windows
    return Popen(
           ^^^^^^
  File "E:\miniforge\envs\builder\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "E:\miniforge\envs\builder\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [WinError 2] The system cannot find the file specified

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "E:\miniforge\envs\builder\Scripts\conda-smithy-script.py", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "E:\miniforge\envs\builder\Lib\site-packages\conda_smithy\cli.py", line 767, in main
    args.subcommand_func(args)
  File "E:\miniforge\envs\builder\Lib\site-packages\conda_smithy\cli.py", line 608, in __call__
    self._call(args, tmpdir)
  File "E:\miniforge\envs\builder\Lib\site-packages\conda_smithy\cli.py", line 613, in _call
    configure_feedstock.main(
  File "E:\miniforge\envs\builder\Lib\site-packages\conda_smithy\configure_feedstock.py", line 2856, in main
    set_exe_file(os.path.join(forge_dir, "build-locally.py"))
  File "E:\miniforge\envs\builder\Lib\site-packages\conda_smithy\feedstock_io.py", line 36, in set_exe_file
    repo.git.execute(["git", "update-index", f"--chmod={mode}", filename])
  File "E:\miniforge\envs\builder\Lib\site-packages\git\cmd.py", line 1275, in execute
    raise GitCommandNotFound(redacted_command, err) from err
git.exc.GitCommandNotFound: Cmd('git') not found due to: FileNotFoundError('[WinError 2] The system cannot find the file specified')
  cmdline: git update-index --chmod=+x E:\conda-forge\boost-feedstock\build-locally.py
```

CC @mgorny

**Comment by h-vetinari** [comment]
Interestingly, I get this error now even after rolling back to conda-smithy 3.44.9 (conda hung on rollback, so I tried recreating the environment). So it might not be related to #2120 after all.

In any case, my workflow is completely broken now - cannot get thing to work despite `--no-check-uptodate` and/or various workarounds like setting `GIT_PYTHON_GIT_EXECUTABLE`.

Here's are the recent environment changes
```
> conda list --revision
[...]
2024-12-14 09:17:30  (rev 61)
     conda-forge-pinning  {2024.12.11.09.08.10 (https://conda.anaconda.org/conda-forge/noarch) -> 2024.12.13.14.31.10 (conda-forge)}
     conda-libmamba-solver  {24.11.1 (https://conda.anaconda.org/conda-forge/noarch) -> 24.9.0 (conda-forge)}
     conda-smithy  {3.44.9 (https://conda.anaconda.org/conda-forge/noarch) -> 3.45.0 (conda-forge)}
     libarchive  {3.7.7 (https://conda.anaconda.org/conda-forge/win-64) -> 3.7.7 (conda-forge)}
     libcurl  {8.10.1 (https://conda.anaconda.org/conda-forge/win-64) -> 8.11.1 (conda-forge)}
     libmamba  {2.0.4 (https://conda.anaconda.org/conda-forge/win-64) -> 1.5.11 (conda-forge)}
     libmambapy  {2.0.4 (https://conda.anaconda.org/conda-forge/win-64) -> 1.5.11 (conda-forge)}
     matplotlib-base  {3.9.3 (https://conda.anaconda.org/conda-forge/win-64) -> 3.9.4 (conda-forge)}
     nodeenv  {1.9.1 (https://conda.anaconda.org/conda-forge/noarch) -> 1.9.1 (conda-forge)}
     python-utils  {3.9.1 (https://conda.anaconda.org/conda-forge/noarch) -> 3.9.1 (conda-forge)}
     rattler-build  {0.32.0 (https://conda.anaconda.org/conda-forge/win-64) -> 0.32.1 (conda-forge)}
     ruff  {0.8.2 (https://conda.anaconda.org/conda-forge/win-64) -> 0.8.3 (conda-forge)}
     simdjson  {3.10.1 (https://conda.anaconda.org/conda-forge/win-64) -> 3.11.3 (conda-forge)}
     spdlog  {1.14.1 (https://conda.anaconda.org/conda-forge/win-64) -> 1.15.0 (conda-forge)}
     uv  {0.5.7 (https://conda.anaconda.org/conda-forge/win-64) -> 0.5.8 (conda-forge)}
    +libgit2-1.8.4 (conda-forge)
    +pygit2-1.16.0 (conda-forge)

2024-12-14 09:41:21  (rev 62)
     conda-smithy  {3.45.0 (https://conda.anaconda.org/conda-forge/noarch) -> 3.44.9 (conda-forge)}

2024-12-14 09:44:47  (rev 63)
    -libgit2-1.8.4 (https://conda.anaconda.org/conda-forge/win-64)
    -pygit2-1.16.0 (https://conda.anaconda.org/conda-forge/win-64)
```
No idea why mamba got downgraded, but that's a sideshow.

I also ran `pixi global upgrade git` (and to debug, downgraded it again to the previous 2.45), and interestingly, there's now one less file
```
>where git
C:\Users\axelo\.pixi\bin\git.bat
```

Always fun when even the roll-back attempts all fail 😑

**Comment by mgorny** [comment]
Hmm, `git` package is GitPython (pygit2 is `pygit2`). Perhaps try uninstalling `pygit2` and `libgit2` and see if that changes anything.

**Comment by h-vetinari** [comment]
That's what I did in 

> ```
> 2024-12-14 09:44:47  (rev 63)
>     -libgit2-1.8.4 (https://conda.anaconda.org/conda-forge/win-64)
>     -pygit2-1.16.0 (https://conda.anaconda.org/conda-forge/win-64)
> ```

But perhaps the removal wasn't clean for some reason...

**Comment by h-vetinari** [comment]
So conda-smithy has dependend on `gitpython` since [forever](https://github.com/conda-forge/conda-smithy-feedstock/blame/5cc6673e3b21311d414323abfe1d83f60e9c6881/recipe/meta.yaml#L41); no idea why that would break now.

**Comment by mgorny** [comment]
Hmm, but unless I'm mistaken, the newest version should not require `GitPython` at all. So perhaps try force-uninstalling it after upgrade and see if that changes anything.

**Comment by h-vetinari** [comment]
If I do that:
```
  File "E:\miniforge\envs\builder2\lib\site-packages\conda_smithy\github.py", line 5, in <module>
    from git import Repo
ModuleNotFoundError: No module named 'git'
```

**Comment by mgorny** [comment]
Now, that looks like a valid bug. I'll take a look at it later today.

**Comment by mgorny** [comment]
For a quick test, could you see if the following patch helps with this immediate error:

```diff
diff --git a/conda_smithy/feedstock_io.py b/conda_smithy/feedstock_io.py
index 900fbdc0..281d3c4f 100644
--- a/conda_smithy/feedstock_io.py
+++ b/conda_smithy/feedstock_io.py
@@ -65,7 +65,7 @@ def write_file(filename):
 
     repo = get_repo(filename)
     if repo:
-        repo.index.add(os.path.relpath(filename, repo.workdir))
+        repo.index.add(os.path.relpath(filename, repo.workdir).replace(os.path.sep, '/'))
         repo.index.write()
 
 ```

My guess is that git expects unix paths. If you confirm, I'll look into preparing a better fix. I'm very sorry about the issue.

**Comment by mgorny** [comment]
Is `git` not in `PATH` for you? It looks as if `subprocess.call(["git"])` didn't work for you.

**Comment by h-vetinari** [comment]
I pointed out in the OP that it's available in the terminal where I use both `git` and `conda smithy`. It might be conceivable that some aspect of the pixi-global-provided-git-from-conda-forge doesn't work though. I checked that the folder where `git.bat` is present is indeed on the PATH, but perhaps what throws the logic is that it's a `.bat` wrapper. That wrapper does
```batch
@echo off
setlocal
@chcp 65001 > nul
@SET "Path=C:\Users\[xxx]\.pixi\envs\git;C:\Users\[xxx]\.pixi\envs\git\Library/mingw-w64/bin;C:\Users\[xxx]\.pixi\envs\git\Library/usr/bin;C:\Users\[xxx]\.pixi\envs\git\Library/bin;C:\Users\[xxx]\.pixi\envs\git\Scripts;C:\Users\[xxx]\.pixi\envs\git\bin;%Path%"
@SET "CONDA_PREFIX=C:\Users\[xxx]\.pixi\envs\git"
@"C:\Users\[xxx]\.pixi\envs\git\Library/bin/git.exe" %*

endlocal
```

Before I got around to checking whether the path of the actual exe was in my path, I just re-updated my `pixi global git` install, and now suddenly, the wrapper is gone:
```>where git
C:\Users\[xxx]\.pixi\bin\git.exe
```

With this unexplained change, I was now also able to reinstall `gitpython` and things keep running (there's some spurious files created upon rerender, but that's a separate issue).

So it looks conceivable that something somewhere somehow corrupted my git install (making it invisible from subprocess, but still visible from my own terminal).

CC @wolfv for awareness
